core/src/main/java/io/jenkins/pluginhealth/scoring/probes/IncrementalBuildDetectionProbe.java

```diff
@@ -0,0 +1,151 @@
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.xml.sax.SAXException;
+
+/**
+ * This probe checks for Incremental Build configuration in plugins.
+ * It looks for configuration in {@code .mvn/maven.config} and {@code .mvn/extensions.xml} files.
+ * The probe is successful when the configurations are found in both the files. Otherwise, it fails.
+ */
+@Component
+@Order(IncrementalBuildDetectionProbe.ORDER)
+public class IncrementalBuildDetectionProbe extends Probe {
+    public static final int ORDER = SCMLinkValidationProbe.ORDER + 100;
+    public static final String KEY = "incremental-build-maven-configuration";
+    private static final Logger LOGGER = LoggerFactory.getLogger(IncrementalBuildDetectionProbe.class);
+    private static final String INCREMENTAL_TOOL = "io.jenkins.tools.incrementals";
+    private static final String INCREMENTAL_TOOL_ARTIFACT_ID = "git-changelist-maven-extension";
+
+    @Override
+    protected ProbeResult doApply(Plugin plugin, ProbeContext context) {
+        final Path scmRepository = context.getScmRepository();
+        final Path mvnConfig = scmRepository.resolve(".mvn");
+        if (Files.notExists(mvnConfig)) {
+            LOGGER.info("Could not find Maven configuration folder {} plugin while running {} probe.", plugin.getName(), key());
+            return ProbeResult.failure(key(), String.format("Could not find Maven configuration folder for the %s plugin.", plugin.getName()));
+        }
+
+        try (Stream<Path> incrementalConfigsStream = Files.find(mvnConfig, 1, (path, $) -> Files.isRegularFile(path) && (path.endsWith("maven.config") || path.endsWith("extensions.xml")))) {
+            List<Path> incrementalConfigsStore = incrementalConfigsStream.toList();
+
+            // Here, we stored the data in a `List`, and we can create a supplier that will create a new `Stream` for us from this data
+            Supplier<Stream<Path>> incrementalConfigs = incrementalConfigsStore::stream;
+
+            Optional<Path> mavenExtensionsFile = incrementalConfigs.get().filter(path -> path.endsWith("extensions.xml")).findFirst();
+            Optional<Path> mavenConfigFile = incrementalConfigs.get().filter(path -> path.endsWith("maven.config")).findFirst();
+
+            if (mavenExtensionsFile.isPresent() && mavenConfigFile.isPresent()) {
+                return isExtensionsXMLConfigured(mavenExtensionsFile.get()) && isMavenConfigConfigured(mavenConfigFile.get())
+                    ? ProbeResult.success(key(), String.format("Incremental Build is configured in the %s plugin.", plugin.getName()))
+                    : ProbeResult.failure(key(), String.format("Incremental Build is not configured in the %s plugin.", plugin.getName()));
+            }
+        } catch (IOException e) {
+            LOGGER.error("Could not read files from .mvn directory for {} plugin while running {} probe.", plugin.getName(), key());
+            return null;
```

Comment:
Is it expected that we return `null`? Isn't it a probe failure? I don't remember what we usually do in that scenario TBH

Reply:
This is the part of the old syntax. When I was using the Supplier the other way. My bad.